### DISCUSSION TITLE
New WebSocket event subscription endpoints

### DISCUSSION THREAD

**Discussion by shramee** [eqlabs/pathfinder#1107] [Polls]
Since v0.5.6 of Pathfinder, an experimental feature to subscribe to real time updates via Websockets has been added.
This feature currently allows getting real-time updates when new headers are appended to the chain.

This discussion is to debate on which new subscription endpoints would be more useful and prioritise their development.

Poll from Telegram,
<img width="358" alt="image" src="https://github.com/eqlabs/pathfinder/assets/11048263/5c5e22b4-f2e6-4f49-abc1-0234fe608b91">

**Comment by franceccon** [comment]
For Starknet React the two main sources of polling calls are blocks and transaction status.

 * **Blocks**: we use this to check for new blocks (latest or pending). This hook is super important because it drives the refresh of other hooks.
 * **Transaction Status**: this data is also refreshed on every block. It would be extremely useful to be able to subscribe to multiple transaction statuses in a single connection.

**Reply by Mirko-von-Leipzig** [reply]
Thanks @fracek.

For transactions:
1. I assume you would just specify the transaction hashes you're interested in? Or is there some more general filter you need, like all transactions from some contract?
2. What would be a sensible lifecycle for the subscription? Once all transactions are `L1_ACCEPTED`? Or `L2_ACCEPTED`? Or should the subscription be kept open indefinitely in case of reorgs etc? i.e. closed at the callers discretion (this probably makes little sense, but just checking).

For blocks:
1. We currently have the block headers as a subscription - is this sufficient? Or do you need block body as well (tx content). How this should interact with pending data is also.. unclear. Currently it only emits complete blocks.

**Reply by franceccon** [reply]
For blocks: the current behaviour is perfect. Block headers is really the only thing needed.


For transactions: yes we would provide the hash. In the starknet react case it's mostly transactions from the user wallet that were submitted by them. Good question about when to close the connection, my opinion is that it should be the clients choice for a number of reasons (for example, how to detect if it disconnected because all tx where accepted or because of a network issue).

**Reply by Mirko-von-Leipzig** [reply]
From a node perspective the concern would be that a client holds the websocket open indefinitely as a form of DoS 🤷 so if the node knows there is no more info to output, it might want to close the socket and move on with life 😀

Maybe one can output a `Complete` notification or something to indicate that 🤔

**Reply by shramee** [reply]
But, the socket could stay open **if** client provides new transactions to start watching?

**Reply by Mirko-von-Leipzig** [reply]
> But, the socket could stay open **if** client provides new transactions to start watching?

Right, the socket yes, I suppose I meant the subscription and not the socket itself.

**Reply by shramee** [reply]
Ahh yeah, Closing the subscription with a `Complete` notif sounds good to me.

**Reply by franceccon** [reply]
I think a `Complete` message is good.

**Comment by Mirko-von-Leipzig** [comment]
Another thing that should be easy to implement is an event subscription. We re-use the event filter as defined in the starknet RPC specification. I've been told this is pretty important for some users.
